tools/16ton/parport/pport-0.6.9/man/pportd.1
.\" Copyright 2004 Corcalciuc V. Horia (rain@sevenrains.ro)
.\" May be distributed under the GNU General Public License
.TH PPORT 1 "12 March 2004"
.SH NAME
pport \- set/toggle parallel port pins on and off
.SH SYNOPSIS
.nf
.BR "pportd [ OPTIONS ]"
.fi
.SH DESCRIPTION
.B pportd 
is the counterpart of the pport program. It reads the input pins of
the parallel port and executes commands based on a configuration file called
.B pportd.conf
that is meant to reside in the sysconfdir of the system.

.B pportd
will read a configuration file, calibrate your parallel port to its internal
functions and then depending on your flags, will fork into the background or
stay in the foreground, being either verbose or silent, executing commands.

.B pportd
will read the input pins of the parallel port also called the status port
and after a relaxation period will execute a configurable command. The PC 
parallel port is a 25 pin D-shaped female connector in the back of the 
computer. It is conventionally used to connect a printer to the computer, 
but also other types of hardware available for this port. Not all 25 pins 
are always needed. 
.B pportd
reads the status pins of the parallel port. Notably pins 10, 11, 12, 13 and
15. Some pins are also simple grounding pins, which you will probably need, 
in order to ground your circuit. Here is a scheme of the parallel port:

.nf

    SLCT|PPR|BSY|ACK|                 
   _____|___|___|___|________________________________________        
 *      |   |   |   |                                         *       
 *     13  12  11  10   9   8   7   6   5   4   3   2   1     *       
  *                                                          *        
   *     25  24  23  22  21  20  19  18  17  16  15  14     *         
    *____|___|___|___|___|___|___|___|____|_______|________*          
         |   |   |   |   |   |   |   |    |       +-+                   
          +---+---+---+---+---+---+---+----+-----+  |                  
                                                 |  +_ERROR_                  
                                                ---                   
                                               ////                   

.fi

.B pportd
uses ACK (Acknowledge), BSY (Busy), PPR (Paper Empty), SLCT (Select) and
ERROR (Error). Pins 17 through 25 are the ground pins of the parallel port.

.SH OPTIONS
.TP
.B \-h | --help
Causes
.B pportd
to pint out a quick help reference.
.TP
.B \-f | --foreground
Hinders
.B pportd
from forking into the background.
.TP
.B \-y | --nosyslog
Prevents
.B pportd
to print to the syslog.
.TP
.B \-r | --resolution
will check the hardware status at the number of milliseconds specified.
Otherwise
.B pportd
defaults to 1ms.
.TP
.B pportd
with no arguments launches the daemon into the background.

.SH HISTORY
This program is an extension of the  uttermost practical Coffee-HOWTO
originally written by Fotis Georgatos, < gef@hellug.gr> and Annie Pinder, <
ajp@leguin.org.uk>. It tries to provide a standard for collecting the
status pins of the parallel port quickly and simply. The "custom" howto
provided, notably Franky-HOWTO, contains various hardware applications
that can be used. It provides an updated set of ideas that could help you
realize the functionality you seek.
.SH AUTHORS
.nf
Written by Corcalciuc V. Horia (rain@sevenrains.ro)
.fi


debian/changelog
linux-4.14.13-16ton-gpd-win-gaming-conf (4.14.13-16ton-gpd-win-gaming-conf-1) xenial; urgency=low

  * Custom built Linux kernel.

 -- Anonymous <root@maddocks-Q552UB>  Thu, 11 Jan 2018 12:07:50 -0500


tools/16ton/parport/pport-0.6.9/src/pportd.c
/*
        pport: Small program to interface your computer with the
               parallel port. Inspired from the Coffee-Machine HOWTO
        
        Copyright (C) 2003  Corcalciuc V. Horia

        This library is free software; you can redistribute it and/or
        modify it under the terms of the GNU Lesser General Public
        License as published by the Free Software Foundation; either
        version 2.1 of the License, or (at your option) any later version.
                
        This library is distributed in the hope that it will be useful,
        but WITHOUT ANY WARRANTY; without even the implied warranty of
        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
        Lesser General Public License for more details.
                                
        You should have received a copy of the GNU Lesser General Public
        License along with this library; if not, write to the Free Software
        Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA      

*/

/* System configuration */

#include "config.h"

/*	Basic includes.	*/

#include <popt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <time.h>
#include <syslog.h>
#include <signal.h>
#include <pthread.h>
#include <sys/wait.h>
#include <sys/io.h>

/*	Our includes.	*/

#include "defines.h"
#include "defext.h"
#include "errorp.h"

struct config
{
  int status[LINEZ];
  char *enable[LINEZ];
  char *disable[LINEZ];
} *cfo;

int
main (const int argc, const char **argv)
{

  enum bool
  {
    false = 0,
    true
  } dofork = true, sylog = true;

  struct timespec *nano;
  const static int pidat[] = { 63, 127, 31, 15, 7 };

  /* Global variables. */

  register int itra = 0;
  int resolution = 1;
  int pid = 0, status = 0;
  int ipin[5], pipin[5], save[5];
  char result;
  volatile int data;
  poptContext optCon;

  /* Variables governing thread creation. */

  pthread_t tida, tidb;

  /* Cmd-line parser. */

  const struct poptOption optionT[] = {
    {"nosyslog", 'y', 0, NULL, 'y',
     "syslog will not be used for logging", NULL},
    {"foreground", 'f', 0, NULL, 'f',
     "daemon will run in the foreground", NULL},
    {"refresh", 'r', POPT_ARG_INT, &resolution, 'r',
     "status check interval in ms", NULL},
    POPT_AUTOHELP {NULL, 0, 0, NULL, 0, NULL}
  };

  optCon = poptGetContext (NULL, argc, argv, optionT, 0);

  while ((result = poptGetNextOpt (optCon)) >= 0)
    switch (result)
      {
      case 'y':
	    sylog = false;
	    break;
      case 'f':
	    dofork = false;
	    break;
      }

  if (result < -1)
    {
      fprintf (stderr, "%s: %s\n",
	       poptBadOption (optCon, POPT_BADOPTION_NOALIAS),
	       poptStrerror (result));
      return 1;
    }

  /* Setup variables accordingly. */

  nano = (struct timespec *) malloc (1 * sizeof (nano));

  nano->tv_sec = 0.001 * resolution;
  nano->tv_nsec = 0;

  /* Get configuration and get the party going. */

  if ((cfo = parse_cf ()) == NULL)
    {
      fprintf (stderr, PARS, PACKAGE_NAME);
      exit (1);
    }

  if (iopl (3))
    {
      fprintf (stderr, DOCS, strerror (errno));
      exit (1);
    }
  else
    {
      fprintf (stdout, "Port 0x%x availible.\n", PBASE);
    }

  fprintf (stdout, "Switch to syslog: ");
  if (sylog == true)
    {
      openlog ("pportd", LOG_PID | LOG_NDELAY, LOG_DAEMON);
      syslog (LOG_INFO, "Starting up...");
      fprintf (stdout, "yes\n");
    }
  else
    fprintf (stdout, "no\n");

  fprintf (stdout, "Switch to daemon mode: ");
  if ((pid = fork ()) < 0)
    {
      fprintf (stderr, "%s", strerror (errno));
      exit (1);
    }
  else if (pid == 0)
    {
      signal (SIGHUP, dhash);
      signal (SIGINT, dhash);
      fprintf (stdout, "done\n");
      while ((data = inb (PBASE + 1)))
	{
	  nanosleep (nano, NULL);
	  for (itra = 0; itra <= 4; itra++)
	    {
	      if (itra == 1)
		ipin[itra] = 0;
	      else
		ipin[itra] = 1;
	    }
	  for (itra = 0; itra <= 4; itra++)
	    {
	      if (data > pidat[itra])
		{
		  if (itra == 1)
		    ipin[itra] = 1;
		  else
		    ipin[itra] = 0;
		  data -= pidat[itra] + 1;
		}
	    }
	  for (itra = 0; itra <= 4; itra++)
	    {
	      if (ipin[itra] == 1 && pipin[itra] != ipin[itra])
		{
		  if (cfo->enable[itra] != NULL)
		    {
		      if (pthread_create
			  (&tida, NULL, (void *) run, cfo->enable[itra]))
			{
			  if (sylog)
			    syslog (LOG_INFO, "%s", strerror (errno));
			  else
			    fprintf (stderr, "%s", strerror (errno));
			  exit (1);
			}
		      if (!dofork)
			fprintf (stdout, "enable: %s\n", cfo->enable[itra]);
		      if (sylog)
			syslog (LOG_INFO, "enable: %s", cfo->enable[itra]);
		    }
		  save[itra] = cfo->status[itra];
		  pipin[itra] = ipin[itra];
		}
	      else if (ipin[itra] == 0)
		{
		  if (cfo->disable[itra] != NULL)
		    {
		      if (pthread_create
			  (&tidb, NULL, (void *) run, cfo->disable[itra]))
			{
			  if (sylog)
			    syslog (LOG_INFO, "%s", strerror (errno));
			  else
			    fprintf (stderr, "%s", strerror (errno));
			  exit (1);
			}
		      if (!dofork)
			fprintf (stdout, "disable: %s\n", cfo->disable[itra]);
		      if (sylog)
			syslog (LOG_INFO, "disable: %s", cfo->disable[itra]);
		    }
		  pipin[itra] = 0;
		  save[itra] = 0;
		}
	    }
	  external_status (save);
	}
    }
  if (!dofork)
    while (wait (&status) != pid);
  return 0;
}

/*	EOF	*/
